linera-chain/src/data_types.rs

```diff
@@ -180,6 +180,25 @@ impl IncomingBundle {
             (message_id, posted_message)
         })
     }
+
+    /// Rearrange the messages in the bundle so that the first message is an `OpenChain` message.
+    /// Returns whether the `OpenChain` message was found at all.
+    pub fn put_openchain_at_front(bundles: &mut [IncomingBundle]) -> bool {
+        let Some(index) = bundles.iter().position(|msg| {
+            matches!(
+                msg.bundle.messages.first(),
+                Some(PostedMessage {
+                    message: Message::System(SystemMessage::OpenChain(_)),
+                    ..
+                })
+            )
+        }) else {
+            return false;
+        };
+
+        bundles.swap(0, index);
```

Comment:
It's dangerous to just swap the messages: It could put messages coming from the same origin into the wrong order.

We'd really need to shift all messages between `0` and `index` one to the right here.

Reply:
Thanks. This was the source of a bug. Fixed in [Fix bug caused by rearranging order of messages in bundle](https://github.com/linera-io/linera-protocol/pull/2745/commits/12380336fe749a8d820fc9418403a09cb67a4ac8).